Userland/Libraries/LibWeb/DOM/Document.cpp

```diff
@@ -1937,15 +1937,40 @@ void Document::set_focused_element(Element* element)
     }
 }
 
-void Document::set_active_element(Element* element)
+// https://html.spec.whatwg.org/multipage/interaction.html#dom-documentorshadowroot-activeelement
+Element const* Document::active_element() const
 {
-    if (m_active_element.ptr() == element)
-        return;
+    // 1. Let candidate be the DOM anchor of the focused area of this DocumentOrShadowRoot's node document.
+    Node const* candidate = focused_element();
 
-    m_active_element = element;
+    // FIXME: What does the spec want us to do when there is no focused area?
+    //        Currently make the document the candidate instead.
+    if (candidate) {
+        // 2. Set candidate to the result of retargeting candidate against this DocumentOrShadowRoot.
+        candidate = retarget<Node const>(candidate, this);
 
-    if (paintable())
-        paintable()->set_needs_display();
+        // 3. If candidate's root is not this DocumentOrShadowRoot, then return null.
+        if (&candidate->root() != this)
+            return nullptr;
+    } else
+        candidate = this;
+
+    // 4. If candidate is not a Document object, then return candidate.
+    if (!is<Document>(candidate))
+        return verify_cast<Element>(candidate);
+
+    auto* candidate_document = verify_cast<Document>(candidate);
```

Comment:
after an `is<>` check, we can use `static_cast` instead of `verify_cast` to skip some if statements and calls to __assert_failed.